cpp/test/test_mp_values.cc

```diff
@@ -14,9 +14,85 @@
 #include <boost/property_tree/ptree.hpp>
 #include <boost/property_tree/info_parser.hpp>
 #include <boost/test/unit_test.hpp>
+#include <boost/math/special_functions/cos_pi.hpp>
 #include <deal.II/grid/grid_generator.h>
 #include <deal.II/grid/tria_accessor.h>
 #include <deal.II/dofs/dof_handler.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_nothing.h>
+#include <deal.II/base/quadrature_lib.h>
+
+BOOST_AUTO_TEST_CASE(fe_values)
+{
+  // MPValues::get_values has recently changed and takes fe_values instead of
+  // cell.  Here we check that if the cell material_id changes, we do not need
+  // to call fe_values.reinit(cell) again.  This is more intended to make sure
+  // we understand how dealii::FEValues behaves than to actually test anything.
+  int constexpr dim = 3;
+  dealii::Triangulation<dim> triangulation;
+  dealii::GridGenerator::hyper_cube(triangulation);
+  dealii::DoFHandler<dim> dof_handler(triangulation);
+  dealii::DoFHandler<dim>::active_cell_iterator cell =
+      dof_handler.begin_active();
+  dealii::FEValues<dim> fe_values(dealii::FE_Nothing<dim>(),
+                                  dealii::QGauss<dim>(0),
+                                  dealii::update_default);
+  fe_values.reinit(cell);
+  BOOST_TEST(cell->material_id() != 255);
+  BOOST_TEST(fe_values.get_cell()->material_id() != 255);
+  cell->set_material_id(255);
+  BOOST_TEST(cell->material_id() == 255);
+  BOOST_TEST(fe_values.get_cell()->material_id() == 255);
+}
+
+// space varying material property built on top of dealii::Function<dim>
+BOOST_AUTO_TEST_CASE(function_space, *boost::unit_test::tolerance(1e-15))
+{
+  int constexpr dim = 2;
+  std::shared_ptr<cap::MPValues<dim>> mp_values;
+
+  // make a dummy triangulation to get an active cell iterator
+  dealii::Triangulation<dim> triangulation;
+  dealii::GridGenerator::hyper_cube(triangulation);
+  dealii::DoFHandler<dim> dof_handler(triangulation);
+  dealii::DoFHandler<dim>::active_cell_iterator cell =
+      dof_handler.begin_active();
+
+  std::vector<dealii::Point<dim>> points = {{0, 0}, {0, 1}, {0.5, 0.5}, {1, 1}};
+  // NOTE: I tried with dealii::update_default but I got a memory access
+  // violation instead of some deal.II exception as I expected...
+  // I was using deal.II in release mode.
```

Comment:
There are very few checks in release mode. That's why it goes at least five times faster.